### PR TITLE
CDFJ+ Support

### PR DESCRIPTION
**Pull request by chrisdwalton**
Add support for new CDFJ+ bankswitching scheme

### CODE REVIEW

File: src/emucore/CartCDF.cxx [hunk 1]
```diff
@@ -194,8 +218,9 @@ inline void CartridgeCDF::callFunction(uInt8 value)
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 uInt8 CartridgeCDF::peek(uInt16 address)
 {
-  address &= 0x0FFF;
+  bool cdfjPlus = (myCDFSubtype == CDFSubtype::CDFJplus);
```
Comment (thrust26) [inline src/emucore/CartCDF.cxx:221]: Why don't you use a method here? you need it for the widget anyway.

File: src/.vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (thrust26) [inline src/.vscode/settings.json:1]: Where are these changes coming from? They are not related to CDFJ+, are they?

File: src/debugger/gui/CartCDFInfoWidget.cxx [hunk 3]
```diff
@@ -23,16 +23,28 @@ CartridgeCDFInfoWidget::CartridgeCDFInfoWidget(
...
+      << "Banks accessible @ $FFF4 to $FFFB\n"
+      << "Startup bank = " << cart.startBank() << "\n";
+
+  } else {
+    size = 8 * 4096;
+    info << describeCDFVersion(cart.myCDFSubtype) << " cartridge\n"
+      << "32K ROM (seven 4K banks are accessible to 2600)\n"
+      << "8K RAM\n"
+      << "Functions accessible @ $FFF0 - $FFF3\n"
+      << "Banks accessible @ $FFF4 to $FFFB\n"
```
Comment (thrust26) [inline src/debugger/gui/CartCDFInfoWidget.cxx:44]: "...$FFF**5** to ... " Right?

File: src/.vscode/settings.json (hunk 2 above)
Comment (chrisdwalton) [reply src/.vscode/settings.json:1]: No - they come from the VS Code IDE - I though I had reverted them, but it looks like it didn't work.

File: src/debugger/gui/CartCDFInfoWidget.cxx (hunk 3 above)
Comment (chrisdwalton) [reply src/debugger/gui/CartCDFInfoWidget.cxx:44]: Actually $FFF4 is also mapped to bank 6 in CDFJ.

File: src/debugger/gui/CartCDFInfoWidget.cxx [hunk 4]
```diff
@@ -23,16 +23,28 @@ CartridgeCDFInfoWidget::CartridgeCDFInfoWidget(
...
-  << "32K ROM, seven 4K banks are accessible to 2600\n"
-  << "8K CDF RAM\n"
-  << "CDF registers accessible @ $FFF0 - $FFF3\n"
-  << "Banks accessible at hotspots $FFF5 to $FFFB\n"
-  << "Startup bank = " << cart.startBank() << "\n";
 
+  if (cart.myCDFSubtype == CartridgeCDF::CDFSubtype::CDFJplus) {
+    size = 512 * 1024;
+    info << describeCDFVersion(cart.myCDFSubtype) << " cartridge\n"
+      << "512K ROM (seven 4K banks are accessible to 2600)\n"
```
Comment (thrust26) [inline src/debugger/gui/CartCDFInfoWidget.cxx:32]: Still only 7 banks? Why 8 hotspots, then?

File: src/debugger/gui/CartCDFWidget.cxx [hunk 5]
```diff
@@ -387,19 +397,20 @@ string CartridgeCDFWidget::bankState()
...
+
   buf << "Bank = " << std::dec << myCart.getBank()
-  << ", hotspot = " << spot[myCart.getBank()];
+  << ", hotspot = " << spot[myCart.getBank() + (isCDFJplus() ? 0 : 1)];
 
   return buf.str();
 }
 
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 uInt32 CartridgeCDFWidget::internalRamSize()
```
Comment (thrust26) [inline src/debugger/gui/CartCDFWidget.cxx:411]: Can the RAM sized be retrieved from the Cart class? That would be better, IMO.

File: src/debugger/gui/CartCDFInfoWidget.cxx [hunk 6]
```diff
@@ -23,16 +23,28 @@ CartridgeCDFInfoWidget::CartridgeCDFInfoWidget(
...
   ostringstream info;
-  info << describeCDFVersion(cart.myCDFSubtype) << " cartridge\n"
-  << "32K ROM, seven 4K banks are accessible to 2600\n"
-  << "8K CDF RAM\n"
-  << "CDF registers accessible @ $FFF0 - $FFF3\n"
-  << "Banks accessible at hotspots $FFF5 to $FFFB\n"
-  << "Startup bank = " << cart.startBank() << "\n";
 
+  if (cart.myCDFSubtype == CartridgeCDF::CDFSubtype::CDFJplus) {
+    size = 512 * 1024;
```
Comment (thrust26) [inline src/debugger/gui/CartCDFInfoWidget.cxx:30]: Can this be provided by Cart class?

File: src/debugger/gui/CartCDFInfoWidget.cxx [hunk 7]
```diff
@@ -23,16 +23,28 @@ CartridgeCDFInfoWidget::CartridgeCDFInfoWidget(
...
+    info << describeCDFVersion(cart.myCDFSubtype) << " cartridge\n"
+      << "512K ROM (seven 4K banks are accessible to 2600)\n"
+      << "32K RAM\n"
```
Comment (thrust26) [inline src/debugger/gui/CartCDFInfoWidget.cxx:33]: Can this be provided by Cart class?

File: src/debugger/gui/CartCDFInfoWidget.cxx [hunk 8]
```diff
@@ -23,16 +23,28 @@ CartridgeCDFInfoWidget::CartridgeCDFInfoWidget(
...
+      << "Functions accessible @ $FFF0 - $FFF3\n"
+      << "Banks accessible @ $FFF4 to $FFFB\n"
+      << "Startup bank = " << cart.startBank() << "\n";
+
+  } else {
+    size = 8 * 4096;
```
Comment (thrust26) [inline src/debugger/gui/CartCDFInfoWidget.cxx:39]: Can this be provided by Cart class?

File: src/debugger/gui/CartCDFInfoWidget.cxx (hunk 4 above)
Comment (chrisdwalton) [reply src/debugger/gui/CartCDFInfoWidget.cxx:32]: The hotspot at $FFFB is being hit at startup on real hardware (not sure why), which means it must be mapped to the startup bank (or removed).    With CDFJ it is rare to need more than 1 or 2 Atari banks, so 7 should be plenty.

File: src/debugger/gui/CartCDFInfoWidget.cxx (hunk 3 above)
Comment (thrust26) [reply src/debugger/gui/CartCDFInfoWidget.cxx:44]: Sorry, I meant to comment the CDFJ+ hotspots. They seem wrong.

Comment (chrisdwalton) [reply src/debugger/gui/CartCDFInfoWidget.cxx:44]: For CDFJ the hotspots are $FFF5 to $FFFB (with bank 6 as startup)
For CDFJ+ they are $FFF4 to $FFFA (with bank 0 as startup)
I would have preferred to keep them the same, but something is hitting $FFFB on startup, meaning it has to map to the startup bank.

Comment (thrust26) [reply src/debugger/gui/CartCDFInfoWidget.cxx:44]: Yup, but in the description the cdfj+ hotspots are copied from cdfj.

**Comment by chrisdwalton** [comment]
Thanks Thomas - I'll refactor and submit a new pull request

**Pull request closed without merging**